File: scripts/fast_github_push.js
const fs = require('fs');
const path = require('path');
const https = require('https');
const crypto = require('crypto');
const { execSync } = require('child_process');

const TOKEN = execSync('/tmp/git-bin/gh auth token').toString().trim();
const OWNER = 'codeswendy-droid';
const REPO = 'tetherstream';
const BRANCH = 'master';

function apiRequest(method, endpoint, body = null) {
  return new Promise((resolve, reject) => {
    const data = body ? JSON.stringify(body) : null;
    const req = https.request({
      hostname: 'api.github.com',
      path: `/repos/${OWNER}/${REPO}${endpoint}`,
      method: method,
      headers: {
        'User-Agent': 'Fast-GitHub-Pusher',
        'Authorization': `Bearer ${TOKEN}`,
        'Accept': 'application/vnd.github.v3+json',
        'Content-Type': 'application/json',
        ...(data ? { 'Content-Length': Buffer.byteLength(data) } : {}),
      },
    }, (res) => {
      let resData = '';
      res.on('data', chunk => resData += chunk);
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            resolve(JSON.parse(resData));
          } catch (e) {
            resolve(resData);
          }
        } else {
          reject(new Error(`API ${method} ${endpoint} failed (${res.statusCode}): ${resData}`));
        }
      });
    });

    req.on('error', reject);
    if (data) req.write(data);
    req.end();
  });
}

function calculateGitBlobSha(contentBuffer) {
  const header = `blob ${contentBuffer.length}\0`;
  const fullBuffer = Buffer.concat([Buffer.from(header, 'utf-8'), contentBuffer]);
  return crypto.createHash('sha1').update(fullBuffer).digest('hex');
}

const IGNORED_NAMES = new Set(['.git', 'node_modules', 'dist', '.next', '.turbo', '.cache', 'tmp', 'package.json.tmp']);

function getAllLocalFiles(dirPath, relativePrefix = '') {
  let fileList = [];
  const items = fs.readdirSync(dirPath);

  for (const item of items) {
    if (IGNORED_NAMES.has(item)) continue;

    const fullPath = path.join(dirPath, item);
    const relPath = relativePrefix ? `${relativePrefix}/${item}` : item;
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      fileList = fileList.concat(getAllLocalFiles(fullPath, relPath));
    } else if (stat.isFile()) {
      const content = fs.readFileSync(fullPath);
      const sha = calculateGitBlobSha(content);
      const mode = (stat.mode & 0o111) ? '100755' : '100644';
      fileList.push({
        path: relPath,
        fullPath: fullPath,
        sha: sha,
        mode: mode,
      });
    }
  }

  return fileList;
}

async function main() {
  console.log(`🚀 Fast Pushing changes to ${OWNER}/${REPO}:${BRANCH}...`);

  // 1. Get current branch head commit
  const refRes = await apiRequest('GET', `/git/ref/heads/${BRANCH}`);
  const parentCommitSha = refRes.object.sha;
  console.log(`📌 Remote Parent Commit SHA: ${parentCommitSha}`);

  // 2. Fetch remote tree
  const remoteCommit = await apiRequest('GET', `/git/commits/${parentCommitSha}`);
  const remoteTreeSha = remoteCommit.tree.sha;
  console.log(`📌 Remote Tree SHA: ${remoteTreeSha}`);

  const remoteTreeRes = await apiRequest('GET', `/git/trees/${remoteTreeSha}?recursive=1`);
  const remoteMap = new Map();
  if (remoteTreeRes.tree) {
    for (const item of remoteTreeRes.tree) {
      if (item.type === 'blob') {
        remoteMap.set(item.path, item.sha);
      }
    }
  }

  // 3. Scan local files
  console.log(`🔍 Scanning local files...`);
  const localFiles = getAllLocalFiles('.');
  console.log(`📁 Found ${localFiles.length} local files.`);

  // 4. Determine modified or new files
  const treeUpdates = [];
  let uploadCount = 0;

  for (const file of localFiles) {
    const remoteSha = remoteMap.get(file.path);

    if (remoteSha !== file.sha) {
      console.log(`⬆️ Uploading changed file: ${file.path}`);
      const content = fs.readFileSync(file.fullPath);
      const blobRes = await apiRequest('POST', '/git/blobs', {
        content: content.toString('base64'),
        encoding: 'base64',
      });

      treeUpdates.push({
        path: file.path,
        mode: file.mode,
        type: 'blob',
        sha: blobRes.sha,
      });
      uploadCount++;
    }
  }

  if (uploadCount === 0) {
    console.log(`✅ Remote is already up to date!`);
    return;
  }

  console.log(`✨ Uploaded ${uploadCount} modified files. Creating new remote tree based on base_tree ${remoteTreeSha}...`);

  // 5. Create new tree using base_tree
  const newTreeRes = await apiRequest('POST', '/git/trees', {
    base_tree: remoteTreeSha,
    tree: treeUpdates,
  });
  console.log(`🌲 Created Tree SHA: ${newTreeRes.sha}`);

  // 6. Create commit
  const commitMsg = 'refactor: complete full-stack rebrand from TetherStream to TitanStream';
  const newCommitRes = await apiRequest('POST', '/git/commits', {
    message: commitMsg,
    tree: newTreeRes.sha,
    parents: [parentCommitSha],
  });
  console.log(`✨ Created Commit SHA: ${newCommitRes.sha}`);

  // 7. Update branch ref
  console.log(`🔄 Updating branch ref ${BRANCH} to ${newCommitRes.sha}...`);
  await apiRequest('PATCH', `/git/refs/heads/${BRANCH}`, {
    sha: newCommitRes.sha,
    force: true,
  });

  console.log(`🎉 SUCCESS! Pushed changes to GitHub repository ${OWNER}/${REPO}:${BRANCH}`);
  console.log(`🔗 Commit URL: https://github.com/${OWNER}/${REPO}/commit/${newCommitRes.sha}`);
}

main().catch(err => {
  console.error('❌ Error during fast push:', err.message);
  process.exit(1);
});

File: scripts/fast_github_push.py
#!/usr/bin/env python3
import os
import sys
import json
import hashlib
import urllib.request
import urllib.parse
import subprocess

def get_gh_token():
    return os.environ.get('GH_TOKEN') or os.environ.get('GITHUB_TOKEN')

TOKEN = get_gh_token()
if not TOKEN:
    print("❌ Error: Could not retrieve GitHub OAuth token.")
    sys.exit(1)

OWNER = os.environ.get('GH_OWNER', 'marlikkodes')
REPO = os.environ.get('GH_REPO', 'titanstream')
BRANCH = os.environ.get('GH_BRANCH', 'main')

def api_request(method, endpoint, body=None):
    url = f"https://api.github.com/repos/{OWNER}/{REPO}{endpoint}"
    data = json.dumps(body).encode('utf-8') if body else None
    
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header('User-Agent', 'Python-Fast-GitHub-Pusher')
    req.add_header('Authorization', f'Bearer {TOKEN}')
    req.add_header('Accept', 'application/vnd.github.v3+json')
    if data:
        req.add_header('Content-Type', 'application/json')
        
    try:
        with urllib.request.urlopen(req) as response:
            res_data = response.read().decode('utf-8')
            return json.loads(res_data) if res_data else {}
    except urllib.error.HTTPError as e:
        err_body = e.read().decode('utf-8')
        raise RuntimeError(f"API {method} {endpoint} failed ({e.code}): {err_body}")

def calculate_git_blob_sha(content_bytes):
    header = f"blob {len(content_bytes)}\0".encode('utf-8')
    full_data = header + content_bytes
    return hashlib.sha1(full_data).hexdigest()

IGNORED_NAMES = {'.git', 'node_modules', 'dist', '.next', '.turbo', '.cache', 'tmp', 'package.json.tmp'}

def get_all_local_files(dir_path, rel_prefix=''):
    file_list = []
    items = os.listdir(dir_path)
    
    for item in items:
        if item in IGNORED_NAMES:
            continue
            
        full_path = os.path.join(dir_path, item)
        rel_path = f"{rel_prefix}/{item}" if rel_prefix else item
        
        if os.path.islink(full_path):
            continue
        elif os.path.isdir(full_path):
            file_list.extend(get_all_local_files(full_path, rel_path))
        elif os.path.isfile(full_path):
            st = os.stat(full_path)
            mode = "100755" if (st.st_mode & 0o111) else "100644"
            with open(full_path, 'rb') as f:
                content = f.read()
            sha = calculate_git_blob_sha(content)
            file_list.append({
                'path': rel_path,
                'full_path': full_path,
                'sha': sha,
                'mode': mode,
                'content': content
            })
            
    return file_list

def main():
    print(f"🚀 Fast Pushing changes to {OWNER}/{REPO}:{BRANCH} via GitHub REST API...")
    
    # 1. Get branch ref
    parent_commit_sha = None
    remote_tree_sha = None
    remote_map = {}
    is_empty = False
    
    try:
        ref_res = api_request('GET', f'/git/ref/heads/{BRANCH}')
        parent_commit_sha = ref_res['object']['sha']
        print(f"📌 Remote Parent Commit SHA: {parent_commit_sha}")
        
        # 2. Get remote tree
        parent_commit = api_request('GET', f'/git/commits/{parent_commit_sha}')
        remote_tree_sha = parent_commit['tree']['sha']
        print(f"📌 Remote Tree SHA: {remote_tree_sha}")
        
        remote_tree_res = api_request('GET', f'/git/trees/{remote_tree_sha}?recursive=1')
        remote_map = {item['path']: item['sha'] for item in remote_tree_res.get('tree', []) if item['type'] == 'blob'}
    except Exception as e:
        print(f"⚠️ Failed to fetch remote reference (might be empty repository): {e}")
        is_empty = True
    
    # 3. Scan local files
    print("🔍 Scanning local workspace files...")
    local_files = get_all_local_files('.')
    print(f"📁 Scanned {len(local_files)} local files.")
    
    # 4. Compare and upload modified/new blobs
    import base64
    tree_updates = []
    upload_count = 0
    
    for file_info in local_files:
        remote_sha = remote_map.get(file_info['path'])
        if remote_sha != file_info['sha']:
            print(f"⬆️ Uploading changed file: {file_info['path']}")
            b64_content = base64.b64encode(file_info['content']).decode('utf-8')
            blob_res = api_request('POST', '/git/blobs', {
                'content': b64_content,
                'encoding': 'base64'
            })
            tree_updates.append({
                'path': file_info['path'],
                'mode': file_info['mode'],
                'type': 'blob',
                'sha': blob_res['sha']
            })
            upload_count += 1
            
    if upload_count == 0:
        print("✅ Remote repository is already completely up to date!")
        return
        
    print(f"✨ Uploaded {upload_count} modified files. Creating new remote tree...")
    
    # 5. Create new tree incrementally in chunks to avoid GitHub API 422 timeouts
    CHUNK_SIZE = 50
    current_base_tree = remote_tree_sha

    for i in range(0, len(tree_updates), CHUNK_SIZE):
        chunk = tree_updates[i:i + CHUNK_SIZE]
        tree_body = {'tree': chunk}
        if current_base_tree:
            tree_body['base_tree'] = current_base_tree
        new_tree_res = api_request('POST', '/git/trees', tree_body)
        current_base_tree = new_tree_res['sha']
        print(f"🌲 Incremental Tree ({i + len(chunk)}/{len(tree_updates)}) SHA: {current_base_tree}")

    new_tree_sha = current_base_tree
    
    # 6. Create commit
    commit_msg = sys.argv[1] if len(sys.argv) > 1 else "refactor: unify Free Trial node into production machine pipeline"
    commit_body = {
        'message': commit_msg,
        'tree': new_tree_sha,
        'parents': [parent_commit_sha] if parent_commit_sha else []
    }
    new_commit_res = api_request('POST', '/git/commits', commit_body)
    new_commit_sha = new_commit_res['sha']
    print(f"✨ Created Remote Commit SHA: {new_commit_sha}")
    
    # 7. Update branch ref
    if is_empty:
        print(f"🔄 Creating branch ref refs/heads/{BRANCH} for new commit {new_commit_sha}...")
        api_request('POST', '/git/refs', {
            'ref': f'refs/heads/{BRANCH}',
            'sha': new_commit_sha
        })
    else:
        print(f"🔄 Updating branch ref {BRANCH} to {new_commit_sha}...")
        api_request('PATCH', f'/git/refs/heads/{BRANCH}', {
            'sha': new_commit_sha,
            'force': True
        })
    
    print(f"🎉 SUCCESS! Successfully pushed branch '{BRANCH}' to GitHub repository {OWNER}/{REPO}!")
    print(f"🔗 Commit URL: https://github.com/{OWNER}/{REPO}/commit/{new_commit_sha}")

if __name__ == '__main__':
    main()
